src/servicex_did_finder_lib/communication.py

```diff
@@ -26,17 +27,52 @@
 __logging.addHandler(logging.NullHandler())
 
 
+class _accumulator:
+    'Track or cache files depending on the mode we are operating in'
+    def __init__(self, sx: ServiceXAdapter, sum: DIDSummary, hold_till_end: bool):
+        self._servicex = sx
+        self._summary = sum
+        self._hold_till_end = hold_till_end
+        self._file_cache: List[Dict[str, Any]] = []
+
+    def add(self, file_info: Dict[str, Any]):
+        'Track and inject the file back into the system'
+        if self._hold_till_end:
+            self._file_cache.append(file_info)
+        else:
+            self._summary.add_file(file_info)
+            if self._summary.file_count == 1:
+                self._servicex.post_transform_start()
+            self._servicex.put_file_add(file_info)
+
+    def send_on(self, count):
+        'Send the accumulated files'
+        if self._hold_till_end:
+            self._hold_till_end = False
+            files = sorted(self._file_cache, key=lambda x: x['paths'])
+            for file_info in files[0:count]:
+                self.add(file_info)
+
+
 async def run_file_fetch_loop(did: str, servicex: ServiceXAdapter, info: Dict[str, Any],
                               user_callback: UserDIDHandler):
-    summary = DIDSummary(did)
     start_time = datetime.now()
-    async for file_info in user_callback(did, info):
 
-        # Track the file, inject back into the system
-        summary.add_file(file_info)
-        if summary.file_count == 1:
-            servicex.post_transform_start()
-        servicex.put_file_add(file_info)
+    summary = DIDSummary(did)
+    did_info = parse_did_uri(did)
+    hold_till_end = did_info.file_count != -1
```

Comment:
This is because you need to make sure repeated queries return the same result - so they have to be sorted before they are returned. Otherwise you inject some randomness into your plots. This was the solution I came up with.

Reply:
The only other approach I was able to think of was that after a query, then re-use the same files in the next query with some indicator. But that requires some real modifications and careful bookkeeping by the physicist as well as SX (though SX already has the info, but it needs to know how to use it).

Reply:
Makes sense - good to make it reproducible